# **Project Introduction: Credit Scoring Analysis**

In this project, the task is to prepare a report for the credit division of a bank focusing on the impact of customers' marital status and number of children on the probability of defaulting on loans. The bank has provided extensive data on customer creditworthiness. Credit assessment is vital for evaluating the ability of borrowers to repay their loans.

**Step of Project**

1. Data Exploration
2. Data Preprocessing
3. Data Categorization:
4. Conclusion

Data Description

1. children: Number of children in the family.
2. days_employed: Duration of employment.
3. dob_years: Age of the customer.
4. education: Level of education.
5. education_id: Identifier for education level.
6. family_status: Marital status.
7. family_status_id: Identifier for marital status.
8. gender: Gender of the customer.
9. income_type: Type of income.
10. debt: Whether the customer has ever defaulted on a loan.
11. total_income: Monthly income.
12. purpose: Purpose of the loan.

This project aims to provide insights into factors influencing credit risk using real-world data that may contain outdated information or inaccuracies, reflecting common challenges in data management scenarios.

## **Data Exploration**

In [1]:
import pandas as pd

In [2]:
dataset = pd.read_csv('/content/credit_scoring_eng.csv')

In [3]:
dataset.shape

(21525, 12)

In [4]:
dataset.head(5)


children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house  
1             car purchase  
2    purchase of the house  
3  supplementary education  
4        to have a wedding

In [5]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Based on the info shown above, all values are filled in the data, so we do not have any problem with it. But we can see an obstacle that might need to be sorted out during this project: There are missing value in the column 'days_employed' and 'total_income', we might will solve the problem by filling the missing value using mean or median of filled value.

Now let us see what are those.

In [6]:
dataset.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

We can see the first column that contains missing value is 'days_employed' the followed by 'total_income'. Then lets see if its semitrical or not by displaying the data with missing value.

In [7]:
dataset.loc[dataset['days_employed'].isna()]

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

We can see the missing value resemble each other between 'days_employed' and 'total_income'. So we can assume that the missing value is symetrical.

In [8]:
dataset_not_null = dataset.dropna()
dataset_not_null

children  days_employed  dob_years            education  education_id  \
0             1   -8437.673028         42    bachelor's degree             0   
1             1   -4024.803754         36  secondary education             1   
2             0   -5623.422610         33  Secondary Education             1   
3             3   -4124.747207         32  secondary education             1   
4             0  340266.072047         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21520         1   -4529.316663         43  secondary education             1   
21521         0  343937.404131         67  secondary education             1   
21522         1   -2113.346888         38  secondary education             1   
21523         3   -3112.481705         38  secondary education             1   
21524         2   -1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                  purpose  
0         40620.102    purchase of the house  
1         17932.802             car purchase  
2         23341.752    purchase of the house  
3         42820.568  supplementary education  
4         25378.572        to have a wedding  
...             ...                      ...  
21520     35966.698     housing transactions  
21521     24959.969        purchase of a car  
21522     14347.610                 property  
21523     39054.888        buying my own car  
21524     13127.587             to buy a car  

[19351 rows x 12 columns]

There is some missing value in column 'days_employed' and 'total_income' thus we will manipulate the data in order to give clear output. we can use methode as bellow:

1. Fill the missing value with median or mean value of filled values in both columns.
2. Median or mean can be from category of certain column that we will find out later.
3. Once missing value in both column are filled, we can continue the project to analyze the data.

In [9]:
dataset_not_null = dataset[dataset['days_employed'].isnull()]

In [10]:
dataset_not_null.describe()

children  days_employed    dob_years  education_id  \
count  2174.000000            0.0  2174.000000   2174.000000   
mean      0.552438            NaN    43.632015      0.800828   
std       1.469356            NaN    12.531481      0.530157   
min      -1.000000            NaN     0.000000      0.000000   
25%       0.000000            NaN    34.000000      0.250000   
50%       0.000000            NaN    43.000000      1.000000   
75%       1.000000            NaN    54.000000      1.000000   
max      20.000000            NaN    73.000000      3.000000   

       family_status_id         debt  total_income  
count       2174.000000  2174.000000           0.0  
mean           0.975161     0.078197           NaN  
std            1.418220     0.268543           NaN  
min            0.000000     0.000000           NaN  
25%            0.000000     0.000000           NaN  
50%            0.000000     0.000000           NaN  
75%            1.000000     0.000000           NaN  
max            4.000000     1.000000           NaN

We can see the missing value are symetrical already, and now lets find out the closest result to fill the missing value.

The possibility of jumbled missing value is low since the data missing is symetrical, but lets find out.

In [11]:
dataset['days_employed'].value_counts(normalize=True)

days_employed
-8437.673028      0.000052
-3507.818775      0.000052
 354500.415854    0.000052
-769.717438       0.000052
-3963.590317      0.000052
                    ...   
-1099.957609      0.000052
-209.984794       0.000052
 398099.392433    0.000052
-1271.038880      0.000052
-1984.507589      0.000052
Name: proportion, Length: 19351, dtype: float64

In [12]:
dataset['days_employed'].describe()

count     19351.000000
mean      63046.497661
std      140827.311974
min      -18388.949901
25%       -2747.423625
50%       -1203.369529
75%        -291.095954
max      401755.400475
Name: days_employed, dtype: float64

In [13]:
401755.400475/365

1100.699727328767

In [14]:
dataset['total_income'].value_counts(normalize=True)

total_income
17312.717    0.000103
42413.096    0.000103
31791.384    0.000103
21005.772    0.000052
18591.443    0.000052
               ...   
27020.895    0.000052
23686.835    0.000052
9606.294     0.000052
28156.762    0.000052
13127.587    0.000052
Name: proportion, Length: 19348, dtype: float64

In [15]:
dataset['family_status'].unique()

array(['married', 'civil partnership', 'widow / widower', 'divorced',
       'unmarried'], dtype=object)

In [16]:
dataset['income_type'].unique()

array(['employee', 'retiree', 'business', 'civil servant', 'unemployed',
       'entrepreneur', 'student', 'paternity / maternity leave'],
      dtype=object)

In [17]:
dataset['dob_years'].unique()

array([42, 36, 33, 32, 53, 27, 43, 50, 35, 41, 40, 65, 54, 56, 26, 48, 24,
       21, 57, 67, 28, 63, 62, 47, 34, 68, 25, 31, 30, 20, 49, 37, 45, 61,
       64, 44, 52, 46, 23, 38, 39, 51,  0, 59, 29, 60, 55, 58, 71, 22, 73,
       66, 69, 19, 72, 70, 74, 75])

In [18]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [19]:
dataset.groupby('education')['total_income'].median()

education
BACHELOR'S DEGREE      26758.7540
Bachelor's Degree      28915.3040
GRADUATE DEGREE        31771.3210
Graduate Degree        15800.3990
PRIMARY EDUCATION      20746.7025
Primary Education      16731.2685
SECONDARY EDUCATION    21488.9210
SOME COLLEGE           26225.8490
Secondary Education    21616.9360
Some College           25664.3970
bachelor's degree      28100.0500
graduate degree        29709.9385
primary education      18770.3230
secondary education    21862.9355
some college           25564.9480
Name: total_income, dtype: float64

The data that missing is from column 'days_employed' and 'total_income', the anomalities from other column may found in the next step that is transformation data

## **Data Processing**

Lets transform the data as method bellow:

1. Make sure the spelling is standardize.
2. Make sure there is no duplicate value.
3. Make sure the wrong input is replaced with better value.
4. Categorize data in order to make easier analysis.

In [20]:
dataset.groupby('education').count()

children  days_employed  dob_years  education_id  \
education                                                               
BACHELOR'S DEGREE         274            251        274           274   
Bachelor's Degree         268            243        268           268   
GRADUATE DEGREE             1              1          1             1   
Graduate Degree             1              1          1             1   
PRIMARY EDUCATION          17             16         17            17   
Primary Education          15             14         15            15   
SECONDARY EDUCATION       772            705        772           772   
SOME COLLEGE               29             22         29            29   
Secondary Education       711            646        711           711   
Some College               47             40         47            47   
bachelor's degree        4718           4222       4718          4718   
graduate degree             4              4          4             4   
primary education         250            231        250           250   
secondary education     13750          12342      13750         13750   
some college              668            613        668           668   

                     family_status  family_status_id  gender  income_type  \
education                                                                   
BACHELOR'S DEGREE              274               274     274          274   
Bachelor's Degree              268               268     268          268   
GRADUATE DEGREE                  1                 1       1            1   
Graduate Degree                  1                 1       1            1   
PRIMARY EDUCATION               17                17      17           17   
Primary Education               15                15      15           15   
SECONDARY EDUCATION            772               772     772          772   
SOME COLLEGE                    29                29      29           29   
Secondary Education            711               711     711          711   
Some College                    47                47      47           47   
bachelor's degree             4718              4718    4718         4718   
graduate degree                  4                 4       4            4   
primary education              250               250     250          250   
secondary education          13750             13750   13750        13750   
some college                   668               668     668          668   

                      debt  total_income  purpose  
education                                          
BACHELOR'S DEGREE      274           251      274  
Bachelor's Degree      268           243      268  
GRADUATE DEGREE          1             1        1  
Graduate Degree          1             1        1  
PRIMARY EDUCATION       17            16       17  
Primary Education       15            14       15  
SECONDARY EDUCATION    772           705      772  
SOME COLLEGE            29            22       29  
Secondary Education    711           646      711  
Some College            47            40       47  
bachelor's degree     4718          4222     4718  
graduate degree          4             4        4  
primary education      250           231      250  
secondary education  13750         12342    13750  
some college           668           613      668

In [21]:
dataset['education'] = dataset['education'].replace({
    "BACHELOR'S DEGREE": "bachelor degree",
    "Bachelor's Degree": "bachelor degree",
    "bachelor's degree": "bachelor degree",
    "GRADUATE DEGREE": "graduate degree",
    "graduate degree": "graduate degree",
    "PRIMARY EDUCATION": "primary education",
    "SECONDARY EDUCATION": "secondary education",
    "SOME COLLEGE": "some college",
    "some college": "some college",
    "primary education": "primary education",
    "secondary education": "secondary education",
    "Secondary Education": "secondary education",
    "Some College": "some college",
    "Primary Education": "primary education",
    "Graduate Degree": "graduate degree"
})

In [22]:
dataset['education'].unique()

array(['bachelor degree', 'secondary education', 'some college',
       'primary education', 'graduate degree'], dtype=object)

In [23]:
dataset['children'].value_counts()

children
 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: count, dtype: int64

There some anomalies in number of children. Above shown that the spread is 0 - 5 with anomalies of value -1 and 20. We can say that because there is no value between 5 to 20 and there is no way the number is bellow 0. There is a possibility -1 and 20 are typos when they are supposed to be 1 and 2.

In [24]:
dataset['children'] = dataset['children'].replace({
    -1:1,
    20:2
})

In [25]:
dataset['children'].value_counts()

children
0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: count, dtype: int64

In [26]:
days_employed_all = dataset['days_employed'].count()
days_employed_missing = dataset['days_employed'].isna().sum()

percentage_missing = (days_employed_missing / days_employed_all) * 100


print(days_employed_all)
print(days_employed_missing)
print(percentage_missing)

19351
2174
11.234561521368406


In this case we will fill missing value in days_employed using the mean value.

In [27]:
dataset.groupby('income_type')['days_employed'].min()

income_type
business                       -17615.563266
civil servant                  -15193.032201
employee                       -18388.949901
entrepreneur                     -520.848083
paternity / maternity leave     -3296.759962
retiree                        328728.720605
student                          -578.751554
unemployed                     337524.466835
Name: days_employed, dtype: float64

In [28]:
median_days_employed = dataset['days_employed'].median()
dataset['days_employed'] = dataset['days_employed'].fillna(median_days_employed)

In [29]:
dataset.isna().sum()

children               0
days_employed          0
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [30]:
dataset.groupby('dob_years').count()

children  days_employed  education  education_id  family_status  \
dob_years                                                                    
0               101            101        101           101            101   
19               14             14         14            14             14   
20               51             51         51            51             51   
21              111            111        111           111            111   
22              183            183        183           183            183   
23              254            254        254           254            254   
24              264            264        264           264            264   
25              357            357        357           357            357   
26              408            408        408           408            408   
27              493            493        493           493            493   
28              503            503        503           503            503   
29              545            545        545           545            545   
30              540            540        540           540            540   
31              560            560        560           560            560   
32              510            510        510           510            510   
33              581            581        581           581            581   
34              603            603        603           603            603   
35              617            617        617           617            617   
36              555            555        555           555            555   
37              537            537        537           537            537   
38              598            598        598           598            598   
39              573            573        573           573            573   
40              609            609        609           609            609   
41              607            607        607           607            607   
42              597            597        597           597            597   
43              513            513        513           513            513   
44              547            547        547           547            547   
45              497            497        497           497            497   
46              475            475        475           475            475   
47              480            480        480           480            480   
48              538            538        538           538            538   
49              508            508        508           508            508   
50              514            514        514           514            514   
51              448            448        448           448            448   
52              484            484        484           484            484   
53              459            459        459           459            459   
54              479            479        479           479            479   
55              443            443        443           443            443   
56              487            487        487           487            487   
57              460            460        460           460            460   
58              461            461        461           461            461   
59              444            444        444           444            444   
60              377            377        377           377            377   
61              355            355        355           355            355   
62              352            352        352           352            352   
63              269            269        269           269            269   
64              265            265        265           265            265   
65              194            194        194           194            194   
66              183            183        183           183            183   
67              167            16

Seems like there is 0 value in the data in column 'dob_years' which is impossible. This may caused by wrong input.

In [31]:
filtered_data = dataset[dataset['dob_years'] == 0]
filtered_data.head(5)

children  days_employed  dob_years            education  education_id  \
99           0  346541.618895          0  secondary education             1   
149          0   -2664.273168          0  secondary education             1   
270          3   -1872.663186          0  secondary education             1   
578          0  397856.565013          0  secondary education             1   
1040         0   -1158.029561          0      bachelor degree             0   

     family_status  family_status_id gender income_type  debt  total_income  \
99         married                 0      F     retiree     0     11406.644   
149       divorced                 3      F    employee     0     11228.230   
270        married                 0      F    employee     0     16346.633   
578        married                 0      F     retiree     0     15619.310   
1040      divorced                 3      F    business     0     48639.062   

                           purpose  
99                             car  
149           housing transactions  
270             housing renovation  
578   construction of own property  
1040                  to own a car

In [32]:
zero_count = dataset['dob_years'].eq(0).sum()
non_zero_count = dataset['dob_years'].ne(0).sum()
total_count = zero_count + non_zero_count

zero_percentage = (zero_count / total_count)
non_zero_percentage = (non_zero_count / total_count)


print(zero_count)
print(non_zero_count)
print(zero_percentage)

101
21424
0.004692218350754936


Based on analysis above, there is no pattern on why the the value in 'dob_years' is 0. And since the percentage is not that much (only 0.4%), we better erase them in order to have clear data

In [33]:
avg_age = round(dataset['dob_years'].mean())
avg_age

43

In [34]:
dataset.loc[dataset['dob_years'] == 0, 'dob_years'] = avg_age

In [35]:
unique_dob_years = dataset['dob_years'].unique()
unique_dob_years.sort()
print(unique_dob_years)

[19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42
 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66
 67 68 69 70 71 72 73 74 75]


In [36]:
dataset['family_status'].unique()

array(['married', 'civil partnership', 'widow / widower', 'divorced',
       'unmarried'], dtype=object)

In [37]:
dataset.groupby('family_status')['children'].mean()

family_status
civil partnership    0.460857
divorced             0.434310
married              0.571648
unmarried            0.232492
widow / widower      0.157292
Name: children, dtype: float64

In [38]:
dataset.groupby('family_status')['total_income'].mean()

family_status
civil partnership    26694.428597
divorced             27189.354550
married              27041.784689
unmarried            26934.069805
widow / widower      22984.208556
Name: total_income, dtype: float64

In [39]:
dataset.groupby('family_status').count()

children  days_employed  dob_years  education  \
family_status                                                      
civil partnership      4177           4177       4177       4177   
divorced               1195           1195       1195       1195   
married               12380          12380      12380      12380   
unmarried              2813           2813       2813       2813   
widow / widower         960            960        960        960   

                   education_id  family_status_id  gender  income_type   debt  \
family_status                                                                   
civil partnership          4177              4177    4177         4177   4177   
divorced                   1195              1195    1195         1195   1195   
married                   12380             12380   12380        12380  12380   
unmarried                  2813              2813    2813         2813   2813   
widow / widower             960               960     960          960    960   

                   total_income  purpose  
family_status                             
civil partnership          3735     4177  
divorced                   1083     1195  
married                   11143    12380  
unmarried                  2525     2813  
widow / widower             865      960

In [40]:
dataset['family_status'] = dataset['family_status'].replace({
   "civil partnership": "married" })

In [41]:
dataset['family_status'].unique()

array(['married', 'widow / widower', 'divorced', 'unmarried'],
      dtype=object)

Actually beside synonym I do not see any problem with "family_status" column.

In [42]:
dataset['gender'].unique()

array(['F', 'M', 'XNA'], dtype=object)

In [43]:
gender_counts = dataset['gender'].value_counts()
gender_counts

gender
F      14236
M       7288
XNA        1
Name: count, dtype: int64

Since its only 1 XNA, we can remove it as well

In [45]:
dataset = dataset[dataset['gender'] != 'XNA']

In [46]:
gender_counts = dataset['gender'].value_counts()
gender_counts

gender
F    14236
M     7288
Name: count, dtype: int64

In [47]:
income = dataset['income_type'].value_counts()
income

income_type
employee                       11119
business                        5084
retiree                         3856
civil servant                   1459
unemployed                         2
entrepreneur                       2
student                            1
paternity / maternity leave        1
Name: count, dtype: int64

In [48]:
filtered_income = dataset[dataset['income_type'].isin(['unemployed', 'student'])]
filtered_income

children  days_employed  dob_years            education  education_id  \
3133          1  337524.466835         31  secondary education             1   
9410          0    -578.751554         22      bachelor degree             0   
14798         0  395302.838654         45      bachelor degree             0   

      family_status  family_status_id gender income_type  debt  total_income  \
3133        married                 0      M  unemployed     1      9593.119   
9410      unmarried                 4      M     student     0     15712.260   
14798       married                 1      F  unemployed     0     32435.602   

                               purpose  
3133   buying property for renting out  
9410      construction of own property  
14798               housing renovation

We can see there is an anomali where source income from unemployment and student. This is unreasonable since those two is not defining where the money come from.

In [49]:
dataset = dataset[~dataset['income_type'].isin(['unemployed', 'student'])]

In [50]:
income = dataset['income_type'].value_counts()
income

income_type
employee                       11119
business                        5084
retiree                         3856
civil servant                   1459
entrepreneur                       2
paternity / maternity leave        1
Name: count, dtype: int64

Now lets see if there is any duplicate values.

In [51]:
duplicate_count = dataset.duplicated().sum()
duplicate_count

71

There is 71 duplicated data in dataset

In [52]:
dataset = dataset.drop_duplicates()

In [53]:
duplicate_count = dataset.duplicated().sum()
duplicate_count

0

In [54]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
Index: 21450 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21450 non-null  int64  
 1   days_employed     21450 non-null  float64
 2   dob_years         21450 non-null  int64  
 3   education         21450 non-null  object 
 4   education_id      21450 non-null  int64  
 5   family_status     21450 non-null  object 
 6   family_status_id  21450 non-null  int64  
 7   gender            21450 non-null  object 
 8   income_type       21450 non-null  object 
 9   debt              21450 non-null  int64  
 10  total_income      19347 non-null  float64
 11  purpose           21450 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.1+ MB


We can see after the cleaning process the value of row is decreased but the data is clear already, you can see different number in total_income because we assumed not all applicant have income. Now lets handle the missing value.

In [55]:
dataset[['days_employed', 'total_income']].describe()

days_employed   total_income
count   21450.000000   19347.000000
mean    56725.044297   26788.435903
std    135082.454333   16476.394122
min    -18388.949901    3306.762000
25%     -2523.921505   16492.834000
50%     -1203.369529   23202.870000
75%      -382.684912   32549.611000
max    401755.400475  362496.645000

### **Handling missing value in total_income**

As stated before there is also missing value in 'total_income' therefore we are going to fill the missing value using mean of filled value for each category. The category are as bellow:

In [56]:
def age_grouping(age):
    if age <= 20:
        value = '11 - 20 years old'
    elif age <= 30:
        value = '21 - 30 years old'
    elif age <= 40:
        value = '31 - 40 years old'
    elif age <= 50:
        value = '41 - 50 years old'
    elif age <= 60:
        value = '51 - 60 years old'
    elif age <= 70:
        value = '61 - 70 years old'
    else:
        value = '>70 years old'
    return value

In [57]:
age_grouping(19)

'11 - 20 years old'

In [58]:
dataset['age_grouping'] = dataset['dob_years'].apply(age_grouping)

In [59]:
dataset['age_grouping'].value_counts()

age_grouping
31 - 40 years old    5731
41 - 50 years old    5360
51 - 60 years old    4518
21 - 30 years old    3650
61 - 70 years old    2022
>70 years old         104
11 - 20 years old      65
Name: count, dtype: int64

Age group is used because the possibility of higher age group earn more money.

In [60]:
filtered_dataset = dataset.dropna(subset=['total_income'])
filtered_dataset.info()

<class 'pandas.core.frame.DataFrame'>
Index: 19347 entries, 0 to 21524
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          19347 non-null  int64  
 1   days_employed     19347 non-null  float64
 2   dob_years         19347 non-null  int64  
 3   education         19347 non-null  object 
 4   education_id      19347 non-null  int64  
 5   family_status     19347 non-null  object 
 6   family_status_id  19347 non-null  int64  
 7   gender            19347 non-null  object 
 8   income_type       19347 non-null  object 
 9   debt              19347 non-null  int64  
 10  total_income      19347 non-null  float64
 11  purpose           19347 non-null  object 
 12  age_grouping      19347 non-null  object 
dtypes: float64(2), int64(5), object(6)
memory usage: 2.1+ MB


In [61]:
filtered_dataset.groupby('age_grouping')['total_income'].mean()

age_grouping
11 - 20 years old    19586.303559
21 - 30 years old    25929.913606
31 - 40 years old    28380.369046
41 - 50 years old    28331.959024
51 - 60 years old    25482.856294
61 - 70 years old    23245.390243
>70 years old        19575.454327
Name: total_income, dtype: float64

In [62]:
age_income = pd.pivot_table(filtered_dataset, index='age_grouping', values='total_income', aggfunc='median').reset_index()
age_income

age_grouping  total_income
0  11 - 20 years old    17257.2770
1  21 - 30 years old    23079.3820
2  31 - 40 years old    24829.1890
3  41 - 50 years old    24563.2635
4  51 - 60 years old    22056.7710
5  61 - 70 years old    19705.8550
6      >70 years old    18611.5935

In [63]:
def replace_income(data, grouping):
    index = 0
    for group in grouping['age_grouping']:
        data.loc[(data['age_grouping'] == group) & (data['total_income'].isnull()), 'total_income'] = grouping.iloc[index, 1]
        index = index + 1
    return data

In [64]:
dataset = replace_income(data=dataset, grouping=age_income)

In [65]:
dataset.isnull().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
age_grouping        0
dtype: int64

In [66]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
Index: 21450 entries, 0 to 21524
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21450 non-null  int64  
 1   days_employed     21450 non-null  float64
 2   dob_years         21450 non-null  int64  
 3   education         21450 non-null  object 
 4   education_id      21450 non-null  int64  
 5   family_status     21450 non-null  object 
 6   family_status_id  21450 non-null  int64  
 7   gender            21450 non-null  object 
 8   income_type       21450 non-null  object 
 9   debt              21450 non-null  int64  
 10  total_income      21450 non-null  float64
 11  purpose           21450 non-null  object 
 12  age_grouping      21450 non-null  object 
dtypes: float64(2), int64(5), object(6)
memory usage: 2.3+ MB


Seems like everything is in order now.

### **Handling missing value in days_employed**

As stated above, missing value also occur in 'days_employed', but the other problem is the value is doesnt make sense since there is negative value and number that doesnt make sense. So these are the next step:

Fill the missing value with mean/median. (its already done but we gave you insight again bellow)
Absolut the negative values.
Fill the anomalies value with our own based determination.

In [67]:
dataset['days_employed'] = abs(dataset['days_employed'])

In [68]:
dataset['years_employed'] = round(dataset['days_employed']/365)

In [69]:
median = pd.pivot_table(dataset, index='age_grouping', values='years_employed', aggfunc='median').reset_index()
median

age_grouping  years_employed
0  11 - 20 years old             2.0
1  21 - 30 years old             3.0
2  31 - 40 years old             4.0
3  41 - 50 years old             5.0
4  51 - 60 years old            12.0
5  61 - 70 years old           960.0
6      >70 years old           983.5

In [70]:
mean = pd.pivot_table(dataset, index='age_grouping', values='years_employed', aggfunc='mean').reset_index()
mean

age_grouping  years_employed
0  11 - 20 years old        1.923077
1  21 - 30 years old        5.380822
2  31 - 40 years old       11.726575
3  41 - 50 years old       43.904478
4  51 - 60 years old      376.746348
5  61 - 70 years old      712.573689
6      >70 years old      832.951923

In [71]:
dataset['years_employed'].describe()

count    21450.000000
mean       165.589231
std        365.648105
min          0.000000
25%          3.000000
50%          5.000000
75%         13.000000
max       1101.000000
Name: years_employed, dtype: float64

In this case we will use median since there is outliers. But the years_employed doesnt make any sense, so we are gonna fix it.

In [72]:
dataset['years_working'] = dataset['dob_years'] - 18
dataset.head(5)

children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42      bachelor degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

  family_status  family_status_id gender income_type  debt  total_income  \
0       married                 0      F    employee     0     40620.102   
1       married                 0      F    employee     0     17932.802   
2       married                 0      M    employee     0     23341.752   
3       married                 0      M    employee     0     42820.568   
4       married                 1      F     retiree     0     25378.572   

                   purpose       age_grouping  years_employed  years_working  
0    purchase of the house  41 - 50 years old            23.0             24  
1             car purchase  31 - 40 years old            11.0             18  
2    purchase of the house  31 - 40 years old            15.0             15  
3  supplementary education  31 - 40 years old            11.0             14  
4        to have a wedding  51 - 60 years old           932.0             35

In [73]:
dataset.loc[dataset['years_employed'] > dataset['years_working'], 'days_employed'] = dataset['years_working'] * 365


In [74]:
dataset['days_employed'].describe()

count    21450.000000
mean      4276.172363
std       5216.342898
min         24.141633
25%       1019.764332
50%       1819.305017
75%       4774.015878
max      20440.000000
Name: days_employed, dtype: float64

In [75]:
dataset['years_employed'] = round(dataset['days_employed']/365)

In [76]:
dataset['years_employed'].describe()

count    21450.000000
mean        11.686247
std         14.308960
min          0.000000
25%          3.000000
50%          5.000000
75%         13.000000
max         56.000000
Name: years_employed, dtype: float64

In [77]:
mean = pd.pivot_table(dataset, index='age_grouping', values='days_employed', aggfunc='mean').reset_index()
mean

age_grouping  days_employed
0  11 - 20 years old     525.577861
1  21 - 30 years old    1270.938715
2  31 - 40 years old    2024.208544
3  41 - 50 years old    2926.285671
4  51 - 60 years old    7087.084840
5  61 - 70 years old   12862.834660
6      >70 years old   16702.663145

In [78]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
Index: 21450 entries, 0 to 21524
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21450 non-null  int64  
 1   days_employed     21450 non-null  float64
 2   dob_years         21450 non-null  int64  
 3   education         21450 non-null  object 
 4   education_id      21450 non-null  int64  
 5   family_status     21450 non-null  object 
 6   family_status_id  21450 non-null  int64  
 7   gender            21450 non-null  object 
 8   income_type       21450 non-null  object 
 9   debt              21450 non-null  int64  
 10  total_income      21450 non-null  float64
 11  purpose           21450 non-null  object 
 12  age_grouping      21450 non-null  object 
 13  years_employed    21450 non-null  float64
 14  years_working     21450 non-null  int64  
dtypes: float64(3), int64(6), object(6)
memory usage: 3.1+ MB


In [79]:
dataset['purpose'].unique()

array(['purchase of the house', 'car purchase', 'supplementary education',
       'to have a wedding', 'housing transactions', 'education',
       'having a wedding', 'purchase of the house for my family',
       'buy real estate', 'buy commercial real estate',
       'buy residential real estate', 'construction of own property',
       'property', 'building a property', 'buying a second-hand car',
       'buying my own car', 'transactions with commercial real estate',
       'building a real estate', 'housing',
       'transactions with my real estate', 'cars', 'to become educated',
       'second-hand car purchase', 'getting an education', 'car',
       'wedding ceremony', 'to get a supplementary education',
       'purchase of my own house', 'real estate transactions',
       'getting higher education', 'to own a car', 'purchase of a car',
       'profile education', 'university education',
       'buying property for renting out', 'to buy a car',
       'housing renovation', 'going

In [80]:
dataset['purpose'] = dataset['purpose'].replace({
    "purchase of the house": "House Ownership",
    "car purchase": "Vehicle Ownership",
    "supplementary education": "Education",
    "having a wedding": "Wedding",
    "housing transactions": "House Ownership",
    "education": "Education",
    "to have a wedding": "Wedding",
    "purchase of the house for my family": "House Ownership",
    "buy real estate": "House Ownership",
    "buy commercial real estate": "House Ownership",
    "buy residential real estate": "House Ownership",
    "construction of own property": "House Construction",
    "property": "House Ownership",
    "building a property": "House Construction",
    "buying a second-hand car": "Vehicle Ownership",
    "buying my own car": "Vehicle Ownership",
    "transactions with commercial real estate": "House Ownership",
    "building a real estate": "House Ownership",
    "housing": "House Ownership",
    "transactions with my real estate": "House Ownership",
    "cars": "Vehicle Ownership",
    "to become educated": "Education",
    "second-hand car purchase": "Vehicle Ownership",
    "getting an education": "Education",
    "car": "Vehicle Ownership",
    "wedding ceremony": "Wedding",
    "to get a supplementary education": "Education",
    "purchase of my own house": "House Ownership",
    "real estate transactions": "House Ownership",
    "getting higher education": "Education",
    "to own a car": "Vehicle Ownership",
    "purchase of a car": "Vehicle Ownership",
    "profile education": "Education",
    "university education": "Education",
    "buying property for renting out": "House Ownership",
    "to buy a car": "Vehicle Ownership",
    "housing renovation": "House Construction",
    "going to university": "Education",
})

In [81]:
dataset['purpose'].unique()

array(['House Ownership', 'Vehicle Ownership', 'Education', 'Wedding',
       'House Construction'], dtype=object)

## **Data Categorization**

The next step is categorize the data, lets explore what we have. In this case we will try purpose first, but theres open possibilities to explore another column.

In [82]:
dataset.groupby('purpose').count()

children  days_employed  dob_years  education  \
purpose                                                             
Education               4013           4013       4013       4013   
House Construction      1859           1859       1859       1859   
House Ownership         8948           8948       8948       8948   
Vehicle Ownership       4306           4306       4306       4306   
Wedding                 2324           2324       2324       2324   

                    education_id  family_status  family_status_id  gender  \
purpose                                                                     
Education                   4013           4013              4013    4013   
House Construction          1859           1859              1859    1859   
House Ownership             8948           8948              8948    8948   
Vehicle Ownership           4306           4306              4306    4306   
Wedding                     2324           2324              2324    2324   

                    income_type  debt  total_income  age_grouping  \
purpose                                                             
Education                  4013  4013          4013          4013   
House Construction         1859  1859          1859          1859   
House Ownership            8948  8948          8948          8948   
Vehicle Ownership          4306  4306          4306          4306   
Wedding                    2324  2324          2324          2324   

                    years_employed  years_working  
purpose                                            
Education                     4013           4013  
House Construction            1859           1859  
House Ownership               8948           8948  
Vehicle Ownership             4306           4306  
Wedding                       2324           2324

In [83]:
dataset['purpose'].unique()

array(['House Ownership', 'Vehicle Ownership', 'Education', 'Wedding',
       'House Construction'], dtype=object)

lets categorize the income level so its easier to analyze.

In [84]:
def income_level(income):
    if income <= 10000:
        return 'Small'
    elif income <= 20000:
        return 'Average'
    elif income <= 25000:
        return 'Above Average'
    elif income <= 50000:
        return 'High'
    elif income > 50000:
        return 'Very High'
    else:
        return 'Unknown'


In [85]:
dataset['income_category'] = dataset['total_income'].apply(income_level)
dataset.head(5)

children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42      bachelor degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0   12775.000000         53  secondary education             1   

  family_status  family_status_id gender income_type  debt  total_income  \
0       married                 0      F    employee     0     40620.102   
1       married                 0      F    employee     0     17932.802   
2       married                 0      M    employee     0     23341.752   
3       married                 0      M    employee     0     42820.568   
4       married                 1      F     retiree     0     25378.572   

             purpose       age_grouping  years_employed  years_working  \
0    House Ownership  41 - 50 years old            23.0             24   
1  Vehicle Ownership  31 - 40 years old            11.0             18   
2    House Ownership  31 - 40 years old            15.0             15   
3          Education  31 - 40 years old            11.0             14   
4            Wedding  51 - 60 years old            35.0             35   

  income_category  
0            High  
1         Average  
2   Above Average  
3            High  
4            High

Let us as well categorize the numeric data.

In [86]:
dataset['purpose'].value_counts()

purpose
House Ownership       8948
Vehicle Ownership     4306
Education             4013
Wedding               2324
House Construction    1859
Name: count, dtype: int64

In [87]:
dataset['age_grouping'].value_counts()

age_grouping
31 - 40 years old    5731
41 - 50 years old    5360
51 - 60 years old    4518
21 - 30 years old    3650
61 - 70 years old    2022
>70 years old         104
11 - 20 years old      65
Name: count, dtype: int64

In [88]:
dataset['children'].value_counts()

children
0    14088
1     4854
2     2128
3      330
4       41
5        9
Name: count, dtype: int64

In [89]:
dataset['debt'].value_counts()

debt
0    19710
1     1740
Name: count, dtype: int64

In [90]:
dataset.describe()

children  days_employed     dob_years  education_id  \
count  21450.000000   21450.000000  21450.000000  21450.000000   
mean       0.480606    4276.172363     43.476084      0.817110   
std        0.756110    5216.342898     12.212740      0.548608   
min        0.000000      24.141633     19.000000      0.000000   
25%        0.000000    1019.764332     33.250000      1.000000   
50%        0.000000    1819.305017     43.000000      1.000000   
75%        1.000000    4774.015878     53.000000      1.000000   
max        5.000000   20440.000000     75.000000      4.000000   

       family_status_id          debt   total_income  years_employed  \
count      21450.000000  21450.000000   21450.000000    21450.000000   
mean           0.973800      0.081119   26449.361525       11.686247   
std            1.421534      0.273024   15690.282306       14.308960   
min            0.000000      0.000000    3306.762000        0.000000   
25%            0.000000      0.000000   17221.291750        3.000000   
50%            0.000000      0.000000   23234.830000        5.000000   
75%            1.000000      0.000000   31327.351000       13.000000   
max            4.000000      1.000000  362496.645000       56.000000   

       years_working  
count   21450.000000  
mean       25.476084  
std        12.212740  
min         1.000000  
25%        15.250000  
50%        25.000000  
75%        35.000000  
max        57.000000

In [91]:
def number_children(children):
    if children == 0:
        return 'Childless'
    elif children == 1:
        return 'Single Child'
    elif children == 2:
        return 'Average'
    elif children <= 4:
        return 'Big Family'
    elif children >= 5:
        return 'Very Big Family'
    else:
        return 'Unknown'

In [92]:
dataset['family_category'] = dataset['children'].apply(number_children)

In [93]:
dataset['family_category'].value_counts()

family_category
Childless          14088
Single Child        4854
Average             2128
Big Family           371
Very Big Family        9
Name: count, dtype: int64

## **Checking Hypothesis**

**Is there a correlation between having children and the probability of loan default?**

In [94]:
df_tmp = pd.crosstab(dataset['children'], dataset['debt'])
df_tmp['bad_rate'] = df_tmp[1]*100/df_tmp.sum(axis=1)
df_tmp

debt          0     1  bad_rate
children                       
0         13025  1063  7.545429
1          4410   444  9.147095
2          1926   202  9.492481
3           303    27  8.181818
4            37     4  9.756098
5             9     0  0.000000

In [95]:
debt_by_children = dataset.loc[dataset['debt'] == 1]
debt_by_children

children  days_employed  dob_years            education  education_id  \
14            0    1844.956182         56      bachelor degree             0   
32            0    4649.910832         34  secondary education             1   
38            0     597.881827         25      bachelor degree             0   
55            0    1203.369529         54  secondary education             1   
75            1    2953.151948         38  secondary education             1   
...         ...            ...        ...                  ...           ...   
21494         0     960.095747         39  secondary education             1   
21511         0     612.569129         29      bachelor degree             0   
21515         1     467.685130         28  secondary education             1   
21522         1    2113.346888         38  secondary education             1   
21523         3    3112.481705         38  secondary education             1   

      family_status  family_status_id gender income_type  debt  total_income  \
14          married                 1      F    business     1     26420.466   
32          married                 1      F    employee     1     22249.194   
38        unmarried                 4      M    employee     1     30759.568   
55          married                 1      F     retiree     1     22056.771   
75          married                 0      M    employee     1     13109.661   
...             ...               ...    ...         ...   ...           ...   
21494       married                 1      M    business     1     28219.135   
21511       married                 1      F    employee     1     22410.956   
21515       married                 0      F    employee     1     17517.812   
21522       married                 1      M    employee     1     14347.610   
21523       married                 0      M    employee     1     39054.888   

                 purpose       age_grouping  years_employed  years_working  \
14       House Ownership  51 - 60 years old             5.0             38   
32               Wedding  31 - 40 years old            13.0             16   
38             Education  21 - 30 years old             2.0              7   
55               Wedding  51 - 60 years old             3.0             36   
75       House Ownership  31 - 40 years old             8.0             20   
...                  ...                ...             ...            ...   
21494          Education  31 - 40 years old             3.0             21   
21511    House Ownership  21 - 30 years old             2.0             11   
21515          Education  21 - 30 years old             1.0             10   
21522    House Ownership  31 - 40 years old             6.0             20   
21523  Vehicle Ownership  31 - 40 years old             9.0             20   

      income_category family_category  
14               High       Childless  
32      Above Average       Childless  
38               High       Childless  
55      Above Average       Childless  
75            Average    Single Child  
...               ...             ...  
21494            High       Childless  
21511   Above Average       Childless  
21515         Average    Single Child  
21522         Average    Single Child  
21523            High      Big Family  

[1740 rows x 17 columns]

In [96]:
debt_by_children['debt'].unique()

array([1])

In [97]:
children_vs_debt = pd.pivot_table(debt_by_children, index=['family_category'], columns='debt', values='children', aggfunc='mean')

total_counts = children_vs_debt.sum().sum()
children_vs_debt_percentage = children_vs_debt / total_counts * 100
children_vs_debt_percentage



debt                     1
family_category           
Average          32.631579
Big Family       51.052632
Childless         0.000000
Single Child     16.315789

In [98]:
debt_by_children_0 = dataset.loc[dataset['debt'] == 0]
debt_by_children_0

children  days_employed  dob_years            education  education_id  \
0             1    8437.673028         42      bachelor degree             0   
1             1    4024.803754         36  secondary education             1   
2             0    5623.422610         33  secondary education             1   
3             3    4124.747207         32  secondary education             1   
4             0   12775.000000         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21518         0   14965.000000         59  secondary education             1   
21519         1    2351.431934         37      graduate degree             4   
21520         1    4529.316663         43  secondary education             1   
21521         0   17885.000000         67  secondary education             1   
21524         2    1984.507589         40  secondary education             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0           married                 0      F    employee     0     40620.102   
1           married                 0      F    employee     0     17932.802   
2           married                 0      M    employee     0     23341.752   
3           married                 0      M    employee     0     42820.568   
4           married                 1      F     retiree     0     25378.572   
...             ...               ...    ...         ...   ...           ...   
21518       married                 0      F     retiree     0     24618.344   
21519      divorced                 3      M    employee     0     18551.846   
21520       married                 1      F    business     0     35966.698   
21521       married                 0      F     retiree     0     24959.969   
21524       married                 0      F    employee     0     13127.587   

                 purpose       age_grouping  years_employed  years_working  \
0        House Ownership  41 - 50 years old            23.0             24   
1      Vehicle Ownership  31 - 40 years old            11.0             18   
2        House Ownership  31 - 40 years old            15.0             15   
3              Education  31 - 40 years old            11.0             14   
4                Wedding  51 - 60 years old            35.0             35   
...                  ...                ...             ...            ...   
21518  Vehicle Ownership  51 - 60 years old            41.0             41   
21519    House Ownership  31 - 40 years old             6.0             19   
21520    House Ownership  41 - 50 years old            12.0             25   
21521  Vehicle Ownership  61 - 70 years old            49.0             49   
21524  Vehicle Ownership  31 - 40 years old             5.0             22   

      income_category family_category  
0                High    Single Child  
1             Average    Single Child  
2       Above Average       Childless  
3                High      Big Family  
4                High       Childless  
...               ...             ...  
21518   Above Average       Childless  
21519         Average    Single Child  
21520            High    Single Child  
21521   Above Average       Childless  
21524         Average         Average  

[19710 rows x 17 columns]

In [99]:
children_vs_debt = pd.pivot_table(debt_by_children_0, index='family_category', columns='debt', values='children', aggfunc='mean')

total_counts = children_vs_debt.sum().sum()
children_vs_debt_percentage = children_vs_debt / total_counts * 100
children_vs_debt_percentage

debt                     0
family_category           
Average          18.003707
Big Family       27.985173
Childless         0.000000
Single Child      9.001853
Very Big Family  45.009267

In [100]:
df_tmp_0 = pd.crosstab(dataset['family_category'], dataset['debt'])
df_tmp_0

debt                 0     1
family_category             
Average           1926   202
Big Family         340    31
Childless        13025  1063
Single Child      4410   444
Very Big Family      9     0

In [101]:
df_tmp_0['bad_rate'] = df_tmp_0[1]*100/df_tmp_0.sum(axis=1)
df_tmp_0

debt                 0     1  bad_rate
family_category                       
Average           1926   202  9.492481
Big Family         340    31  8.355795
Childless        13025  1063  7.545429
Single Child      4410   444  9.147095
Very Big Family      9     0  0.000000

Conclusion

Based on calculation we can see the the highest bad debt happened to family that having big number of family member with staggering 51% of bad debt case. Which is make sense since bigger family needs bigger home, bigger car, as well another daily needs.

But based on the percentage within category, the highest probability for bad debt is average family with 9.4%


**Is there a correlation between marital status and the probability of loan default?**

In [102]:
debt_by_family = dataset.loc[dataset['debt'] == 1]
debt_by_family

children  days_employed  dob_years            education  education_id  \
14            0    1844.956182         56      bachelor degree             0   
32            0    4649.910832         34  secondary education             1   
38            0     597.881827         25      bachelor degree             0   
55            0    1203.369529         54  secondary education             1   
75            1    2953.151948         38  secondary education             1   
...         ...            ...        ...                  ...           ...   
21494         0     960.095747         39  secondary education             1   
21511         0     612.569129         29      bachelor degree             0   
21515         1     467.685130         28  secondary education             1   
21522         1    2113.346888         38  secondary education             1   
21523         3    3112.481705         38  secondary education             1   

      family_status  family_status_id gender income_type  debt  total_income  \
14          married                 1      F    business     1     26420.466   
32          married                 1      F    employee     1     22249.194   
38        unmarried                 4      M    employee     1     30759.568   
55          married                 1      F     retiree     1     22056.771   
75          married                 0      M    employee     1     13109.661   
...             ...               ...    ...         ...   ...           ...   
21494       married                 1      M    business     1     28219.135   
21511       married                 1      F    employee     1     22410.956   
21515       married                 0      F    employee     1     17517.812   
21522       married                 1      M    employee     1     14347.610   
21523       married                 0      M    employee     1     39054.888   

                 purpose       age_grouping  years_employed  years_working  \
14       House Ownership  51 - 60 years old             5.0             38   
32               Wedding  31 - 40 years old            13.0             16   
38             Education  21 - 30 years old             2.0              7   
55               Wedding  51 - 60 years old             3.0             36   
75       House Ownership  31 - 40 years old             8.0             20   
...                  ...                ...             ...            ...   
21494          Education  31 - 40 years old             3.0             21   
21511    House Ownership  21 - 30 years old             2.0             11   
21515          Education  21 - 30 years old             1.0             10   
21522    House Ownership  31 - 40 years old             6.0             20   
21523  Vehicle Ownership  31 - 40 years old             9.0             20   

      income_category family_category  
14               High       Childless  
32      Above Average       Childless  
38               High       Childless  
55      Above Average       Childless  
75            Average    Single Child  
...               ...             ...  
21494            High       Childless  
21511   Above Average       Childless  
21515         Average    Single Child  
21522         Average    Single Child  
21523            High      Big Family  

[1740 rows x 17 columns]

In [103]:
family_vs_debt = pd.pivot_table(debt_by_family, index=['family_status'], columns='debt', values='children', aggfunc='mean')

family_total_counts = family_vs_debt.sum().sum()
family_vs_debt_percentage = family_vs_debt / family_total_counts * 100
family_vs_debt_percentage

debt                     1
family_status             
divorced         29.744488
married          38.988863
unmarried        18.223927
widow / widower  13.042722

In [104]:
family_vs_debt = pd.pivot_table(debt_by_family, index=['family_status', 'purpose'], columns='debt', values='children', aggfunc='mean')

family_total_counts = family_vs_debt.sum().sum()
family_vs_debt_percentage = family_vs_debt / family_total_counts * 100
family_vs_debt_percentage

debt                                        1
family_status   purpose                      
divorced        Education            4.862614
                House Construction   9.184938
                House Ownership      6.298243
                Vehicle Ownership    6.560670
married         Education            8.735675
                House Construction  10.113203
                House Ownership      8.146392
                Vehicle Ownership    9.250545
                Wedding              7.110920
unmarried       Education            2.666595
                House Construction   3.936402
                House Ownership      4.803133
                Vehicle Ownership    3.864395
widow / widower Education            2.755481
                House Construction   6.888703
                House Ownership      3.444352
                Vehicle Ownership    1.377741

In [105]:
df_tmp = pd.crosstab(dataset['family_status'], dataset['debt'])
df_tmp.head()

debt                 0     1
family_status               
divorced          1110    85
married          15169  1318
unmarried         2535   274
widow / widower    896    63

In [106]:
df_tmp['bad_rate'] = df_tmp[1]*100/df_tmp.sum(axis=1)
df_tmp

debt                 0     1  bad_rate
family_status                         
divorced          1110    85  7.112971
married          15169  1318  7.994177
unmarried         2535   274  9.754361
widow / widower    896    63  6.569343

Conclusion

Based on calculation we can see that married couple has highest number percentage of bad debt with 38.9% out of all category. This is make sense when become a family they needs home & car of their own. But if we see spesifically by status category among its success and failed, we can see that unmarried couple has the highest percentage of possibility of bad debt with 9.7& possibility and followed by married with 7.9%.

So by total loan given the highest percentage of bad debt is married couple with 38.9% of total bad debt, but if we calculate the percentage within category unmarried couple has the highest percentage with 9.7%


**Is there a correlation between income level and the probability of loan default?**

In [107]:
debt_earning = dataset.loc[dataset['debt'] == 1]

In [108]:
earning_vs_debt = pd.pivot_table(debt_earning, index='income_category', columns='debt', values='children', aggfunc='mean')

earning_total_counts = earning_vs_debt.sum().sum()
earning_vs_debt_percentage = earning_vs_debt / earning_total_counts * 100
earning_vs_debt_percentage

debt                     1
income_category           
Above Average    20.671958
Average          18.914171
High             19.287153
Small            20.297914
Very High        20.828804

In [109]:
df_tmp_2 = pd.crosstab(dataset['income_category'], dataset['debt'])
df_tmp_2.head()

debt                0    1
income_category           
Above Average    4821  446
Average          6095  562
High             6698  583
Small             868   57
Very High        1228   92

In [110]:
df_tmp_2['bad_rate'] = df_tmp_2[1]*100/df_tmp_2.sum(axis=1)
df_tmp_2

debt                0    1  bad_rate
income_category                     
Above Average    4821  446  8.467818
Average          6095  562  8.442241
High             6698  583  8.007142
Small             868   57  6.162162
Very High        1228   92  6.969697

Conclusion

Based on the calculation its actually vary if we see the cause of bad debt using income category which only have different approximately 1% from each category. But the highest is when the income categorize as very high with range above 50000. And if we calculate percentage within each income type the percentage shows the same range possibility, but the highest possibility land on above average and average with bracket of 10000 to 25000 with percentage of 8.6% and 8.4%.

So by percentage of total loan the category of income very high has the vast amount of loan with 20.8% in total but by its possibility family with above average and average income has more possibility with 8.6% and 8.4% of bad debt possibility.

**How does the loan purpose affect the default rate?**

In [111]:
debt_purpose = dataset.loc[dataset['debt'] == 1]

In [112]:
purpose_vs_debt = pd.pivot_table(debt_purpose, index='purpose', columns='debt', values='children', aggfunc='mean')
purpose_total_counts = purpose_vs_debt.sum().sum()
purpose_vs_debt_percentage = purpose_vs_debt / purpose_total_counts * 100
purpose_vs_debt_percentage

debt                        1
purpose                      
Education           19.063030
House Construction  23.349910
House Ownership     19.100438
Vehicle Ownership   19.913032
Wedding             18.573591

In [113]:
df_tmp_3 = pd.crosstab(dataset['purpose'], dataset['debt'])
df_tmp_3.head()

debt                   0    1
purpose                      
Education           3643  370
House Construction  1728  131
House Ownership     8298  650
Vehicle Ownership   3903  403
Wedding             2138  186

In [114]:
df_tmp_3['bad_rate'] = df_tmp_3[1]*100/df_tmp_3.sum(axis=1)
df_tmp_3

debt                   0    1  bad_rate
purpose                                
Education           3643  370  9.220035
House Construction  1728  131  7.046799
House Ownership     8298  650  7.264193
Vehicle Ownership   3903  403  9.359034
Wedding             2138  186  8.003442

Conclusion

Based on the calculation we can see the highest possibility is home construction with 23.3% and followed by education, home ownership, vehicle ownership, and wedding with the range from 18.5% to 19.91%. Home construction is at risk because the estimated cost can overrun then if its use for productive asset cost overrun will affect the repayment of debt.

By calculate the percentage within each category of purpose, we discovered new fact that the homw ownership and home construction placed bottom, the highest still car ownership with 9.3% and education with 9.2%.

In [140]:
debt_all = dataset.loc[dataset['debt'] == 1]

In [141]:
purpose_children_vs_debt = pd.pivot_table(debt_all, index=['family_category','purpose'], columns='debt', values='children', aggfunc='mean')
purpose_children_total_counts = purpose_children_vs_debt.sum().sum()
purpose_children_vs_debt_percentage = purpose_children_vs_debt / purpose_children_total_counts * 100
purpose_children_vs_debt_percentage

debt                                        1
family_category purpose                      
Average         Education            6.575342
                House Construction   6.575342
                House Ownership      6.575342
                Vehicle Ownership    6.575342
                Wedding              6.575342
Big Family      Education            9.863014
                House Construction   9.863014
                House Ownership     10.684932
                Vehicle Ownership   10.410959
                Wedding              9.863014
Childless       Education            0.000000
                House Construction   0.000000
                House Ownership      0.000000
                Vehicle Ownership    0.000000
                Wedding              0.000000
Single Child    Education            3.287671
                House Construction   3.287671
                House Ownership      3.287671
                Vehicle Ownership    3.287671
                Wedding              3.287671

In [142]:
purpose_children_vs_debt = pd.pivot_table(dataset, index=['family_category', 'purpose'], columns='debt', values='children', aggfunc='count').reset_index()
purpose_children_vs_debt

debt  family_category             purpose       0      1
0             Average           Education   371.0   47.0
1             Average  House Construction   173.0   18.0
2             Average     House Ownership   808.0   72.0
3             Average   Vehicle Ownership   365.0   50.0
4             Average             Wedding   209.0   15.0
5          Big Family           Education    72.0    4.0
6          Big Family  House Construction    28.0    4.0
7          Big Family     House Ownership   146.0   12.0
8          Big Family   Vehicle Ownership    64.0    6.0
9          Big Family             Wedding    30.0    5.0
10          Childless           Education  2413.0  229.0
11          Childless  House Construction  1155.0   72.0
12          Childless     House Ownership  5440.0  404.0
13          Childless   Vehicle Ownership  2602.0  243.0
14          Childless             Wedding  1415.0  115.0
15       Single Child           Education   786.0   90.0
16       Single Child  House Construction   372.0   37.0
17       Single Child     House Ownership  1899.0  162.0
18       Single Child   Vehicle Ownership   870.0  104.0
19       Single Child             Wedding   483.0   51.0
20    Very Big Family           Education     1.0    NaN
21    Very Big Family     House Ownership     5.0    NaN
22    Very Big Family   Vehicle Ownership     2.0    NaN
23    Very Big Family             Wedding     1.0    NaN

In [152]:
# Convert the 'debt' columns to numeric type
purpose_children_vs_debt[1] = pd.to_numeric(purpose_children_vs_debt[1], errors='coerce')
purpose_children_vs_debt[0] = pd.to_numeric(purpose_children_vs_debt[0], errors='coerce')

# Calculate the 'bad_rate' excluding non-numeric columns
purpose_children_vs_debt['bad_rate'] = purpose_children_vs_debt[1] * 100 / (purpose_children_vs_debt[0] + purpose_children_vs_debt[1])

purpose_children_vs_debt.sort_values('bad_rate', ascending=False)

debt  family_category             purpose       0      1   bad_rate
9          Big Family             Wedding    30.0    5.0  14.285714
6          Big Family  House Construction    28.0    4.0  12.500000
3             Average   Vehicle Ownership   365.0   50.0  12.048193
0             Average           Education   371.0   47.0  11.244019
18       Single Child   Vehicle Ownership   870.0  104.0  10.677618
15       Single Child           Education   786.0   90.0  10.273973
19       Single Child             Wedding   483.0   51.0   9.550562
1             Average  House Construction   173.0   18.0   9.424084
16       Single Child  House Construction   372.0   37.0   9.046455
10          Childless           Education  2413.0  229.0   8.667676
8          Big Family   Vehicle Ownership    64.0    6.0   8.571429
13          Childless   Vehicle Ownership  2602.0  243.0   8.541301
2             Average     House Ownership   808.0   72.0   8.181818
17       Single Child     House Ownership  1899.0  162.0   7.860262
7          Big Family     House Ownership   146.0   12.0   7.594937
14          Childless             Wedding  1415.0  115.0   7.516340
12          Childless     House Ownership  5440.0  404.0   6.913073
4             Average             Wedding   209.0   15.0   6.696429
11          Childless  House Construction  1155.0   72.0   5.867971
5          Big Family           Education    72.0    4.0   5.263158
20    Very Big Family           Education     1.0    0.0   0.000000
21    Very Big Family     House Ownership     5.0    0.0   0.000000
22    Very Big Family   Vehicle Ownership     2.0    0.0   0.000000
23    Very Big Family             Wedding     1.0    0.0   0.000000

Based on the data, significant variations in default rates are observed across different family categories and loan purposes. Big families show the highest default rates, particularly for weddings (14.29%) and house construction (12.50%), suggesting higher financial strain in these contexts. Average families also exhibit notable default risks, notably for vehicle ownership (12.05%) and education (11.24%) loans. Single-child families display moderate default rates ranging from 7.86% to 10.68%, while childless families generally show lower default rates, ranging from 5.87% to 8.67%. Interestingly, very big families show no defaults across all recorded purposes, which may reflect a smaller sample size or unique financial stability. These findings underline the importance of family dynamics and loan purpose in assessing credit risk and informing lending decisions.

## **Summary**

In this project given the data of historical performance of creditors and its following information. But during data understanding we found that:

1. Missing value found in column 'days_employed' and 'total_income'.
2. In 'days_employed' column has unrealistic value.
3. Redundant in column 'purpose'.
4. Unrealistic value in column 'children'.
5. Uncategorized data.

So some adjustmen has been made.

1. Using average to filled missing value in 'days_employed' and 'total_income'. for column 'days_employed' we also using absolut since there is negation value and also using realistic value to fill unrealistic value. Such as we substract the column 'dob_years' with 18 (the start age of legally working) in order to replace the unrealistic value. And for column 'total_income' we categrized first based on 'age_category' so the result is closed to relevant, then filled the missing value with the average value from each age category.
2. Redundant value has been replaced in column purpose with replace function.
3. Unrealistic value in column children has been replaced with realistic value (-1) to (1) and (20) to (2). Since there is no way having negation children and spread is only until 5 for top value.
4. Categorized data in column 'income', 'children', 'days_employed'.

The conclusion we can give from the data above is

1. Big families demonstrate the highest default rates, notably for weddings (14.29%) and house construction (12.50%), indicating higher financial strain in these scenarios.
2. Average families show significant default risks, particularly for vehicle ownership (12.05%) and education (11.24%) loans.
3. Single-child families exhibit moderate default rates ranging from 7.86% to 10.68% across different loan purposes.
4. Childless families generally exhibit lower default rates, ranging from 5.87% to 8.67%.
5. Very big families show no defaults across all recorded purposes, possibly due to a smaller sample size or unique financial stability.

These findings underscore the critical role of family dynamics and loan purpose in evaluating credit risk and informing prudent lending practices.